actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -248,19 +250,28 @@ def verify_authenticity_token # :doc:
         mark_for_same_origin_verification!
 
         if !verified_request?
-          if logger && log_warning_on_csrf_failure
-            if valid_request_origin?
-              logger.warn "Can't verify CSRF token authenticity."
-            else
-              logger.warn "HTTP Origin header (#{request.origin}) didn't match request.base_url (#{request.base_url})"
-            end
-          end
+          logger.warn unverified_request_warning_message if logger && log_warning_on_csrf_failure
+
           handle_unverified_request
         end
       end
 
       def handle_unverified_request # :doc:
-        forgery_protection_strategy.new(self).handle_unverified_request
+        protection_strategy = forgery_protection_strategy.new(self)
+
+        if protection_strategy.is_a?(ProtectionMethods::Exception)
```

Comment:
Instead of having to know what a `ProtectionMethods::Exception` is and what it's interface is, we could also check if it responds to `:warning_message`.
It also makes the code a bit clearer: we can set `warning_message` if it's available.
```suggestion
        if protection_strategy.respond_to?(:warning_message)
```

Reply:
Good idea! https://github.com/rails/rails/pull/41822/commits/7b638685883df5f476395500ce86c96dc8b342f6